packages/attune/src/server/mod.rs

```diff
@@ -88,16 +97,81 @@ pub async fn new(state: ServerState, default_api_token: Option<String>) -> Route
             "/packages",
             post(pkg::upload::handler.layer(DefaultBodyLimit::disable())),
         )
-        .route("/packages/{package_sha256sum}", get(pkg::info::handler));
+        .route("/packages/{package_sha256sum}", get(pkg::info::handler))
+        .route("/_test/error", get(test_error))
+        .route("/_test/panic", get(test_panic));
+
+    // The intention of error handling middleware here is that:
+    // - `handle_non_success` handles responses from handlers and axum itself, converting errors to `ErrorResponse`.
+    // - `handle_middleware_error` handles errors from the middleware stack, converting them to `ErrorResponse`.
+    // - `handle_panic` handles panics, converting them to `ErrorResponse`.
     Router::new()
         .nest("/api/v0", api)
-        .layer(TraceLayer::new_for_http())
-        // FIXME: Use a custom ResponseForPanic so that the response body is a
-        // valid `api::ErrorResponse` on 500, which is what all the CLI parsing
-        // logic expects on a non-200 response.
-        //
-        // FIXME: We also need to return valid `api::ErrorResponse` on 404s,
-        // 405s, etc.
-        .layer(CatchPanicLayer::new())
+        .layer(axum::middleware::from_fn(handle_non_success))
+        .layer(
+            ServiceBuilder::new()
+                .layer(TraceLayer::new_for_http())
+                .layer(CatchPanicLayer::custom(handle_panic))
+                .layer(HandleErrorLayer::new(handle_middleware_error))
+                .timeout(Duration::from_secs(600)),
+        )
         .with_state(state)
 }
+
+async fn handle_non_success(request: Request, next: Next) -> Response {
+    let uri = request.uri().to_string();
+    let response = next.run(request).await;
+    let status = response.status();
+    match status {
+        StatusCode::NOT_FOUND => ErrorResponse::new(
```

Comment:
We have a couple handlers that intentionally return informative 404s etc. for the client. For example, if you try to edit a repository that doesn't exist, you get an informative 404 where the client uses the body to display a message to the user. How does this logic interact with those handlers?

Is there a way we can specially handle "route doesn't exist"? If not, then maybe this is better handled on the client side.

Alternatively, I wonder if we can use a heuristic like "if the body parses into an `ErrorResponse`, proxy it through. Or maybe even something cheaper like "if the body is length zero, proxy it through".

Reply:
So long as we aren't ever sending big bodies we can definitely buffer and inspect the response to handle this case.

if we ever need to start sending big bodies there are other things we can do to handle it too. will add this in.

Reply:
@elldritch updated, ref https://github.com/attunehq/attune/pull/67/commits/9157d2d86baaa6fb6a5266e458e1f8b821ed867f and the changes to the PR description.